### PR TITLE
[model] feat: add Dual Chunk Attention

### PR DESCRIPTION
**Pull request by Ternura143**
# What does this PR do ?

Add Dual Chunk Attention (DCA) support for RoPE- and YARN-based GPT training in Megatron Bridge.

This is the Bridge-side integration of the DCA implementation developed in the companion draft PR NVIDIA/Megatron-LM#4048 for NVIDIA/Megatron-LM#2797. Since that MCore change is not part of Bridge's pinned MCore revision, the generic DCA implementation lives under `megatron.bridge.models.transformer`, while GPT-specific builder/provider and layer-spec wiring lives under `megatron.bridge.models.gpt`.

# Changelog

- Add generic DCA attention modules and a dedicated transformer configuration.
- Add modern GPT builder and legacy GPT provider integration.
- Preserve the existing MCore GPT decoder block spec and replace only self-attention/core-attention with DCA.
- Merge chunk outputs using differentiable log-sum-exp normalization.
- Preserve LSE-derived Q/K gradients in the FlashAttention path, with an unfused PyTorch fallback.
- Keep FlashAttention optional through a guarded import; no dependency changes are required.
- Add explicit validation for unsupported training combinations and incomplete YARN configuration.
- Add CPU, CUDA, MHA, GQA, RoPE, YARN, builder/provider, and training-config tests.
- Add DCA configuration, constraints, and correctness documentation.

# Validation

- A800 CUDA DCA/GPT unit tests: 34 passed.
- CPU DCA/GPT unit tests: 31 passed, 3 CUDA-only tests skipped.
- Training cross-config validation tests: 2 passed.
- FlashAttention-vs-unfused output and Q/K/V gradient parity passed for MHA and GQA.
- Legacy provider RoPE GPT forward/backward smoke test passed with Transformer Engine and selective core-attention recompute.
- Modern builder RoPE and YARN GPT forward/backward smoke tests passed with Transformer Engine and selective core-attention recompute.
- `pre-commit run --all-files`: passed.
- `git diff --check`: passed.

# GitHub Actions CI

An NVIDIA developer will need to approve and trigger CI for the updated commit.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (FlashAttention remains optional and has an unfused fallback.)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

# Additional Information

- Related Megatron-LM draft PR: NVIDIA/Megatron-LM#4048
- Related Megatron-LM issue: NVIDIA/Megatron-LM#2797

### CODE REVIEW

File: src/megatron/bridge/models/gpt/dca_attention.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
+# Copyright (c) 2026, NVIDIA CORPORATION.  All rights reserved.
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/gpt/dca_attention.py]: structure-wise, can we move this to "bridge/models/transformer" keep in sync with MCore's side of naming.

Comment (yaoyu-33) [inline src/megatron/bridge/models/gpt/dca_attention.py]: Thanks for contribution. For this change we may need brief doc or example, is there any experiment done to prove the correctness of implementation?

Comment (Ternura143) [reply src/megatron/bridge/models/gpt/dca_attention.py]: Thanks, this makes sense. I’ll move the generic DCA attention implementation under `src/megatron/bridge/models/transformer/` and keep only the GPT-specific layer-spec/config wiring under `models/gpt` as the initial integration point.

Comment (Ternura143) [reply src/megatron/bridge/models/gpt/dca_attention.py]: I’ll submit a new commit asap and include a brief document or example before requesting a re-review. Sry for the delay :)

**Comment by yaoyu-33** [comment]
Could you please check the merge in `_merge_chunk_attention_outputs()`? The LSE-derived merge weights are currently detached before normalization. Those weights depend on Q/K and are part of the normalized attention merge, so detaching them may change the training gradients relative to the intended attention computation.

Please remove the detach or justify it with a gradient-parity test.

File: src/megatron/bridge/models/gpt/dca_attention.py (hunk 1 above)
Comment (Ternura143) [reply src/megatron/bridge/models/gpt/dca_attention.py]: > structure-wise, can we move this to "bridge/models/transformer" keep in sync with MCore's side of naming.

Implemented in the latest update.

Comment (Ternura143) [reply src/megatron/bridge/models/gpt/dca_attention.py]: > Thanks for contribution. For this change we may need brief doc or example, is there any experiment done to prove the correctness of implementation?

Done in the latest update. I added a usage and constraints section to the attention-optimization docs, including both RoPE and YARN configuration.

For correctness, the tests compare long-sequence outputs and Q/K/V gradients against a direct token-level DCA implementation, verify exact float64 LSE partition merging, short-sequence causal-attention parity, causality, and FlashAttention-vs-unfused output/Q/K/V gradient parity for both MHA and GQA. The updated CUDA target has 34 passing tests. I also ran full GPT forward/backward smoke tests through both the legacy provider and modern builder paths with Transformer Engine and selective core-attention recompute, including a modern YARN configuration.

**Comment by Ternura143** [comment]
> Could you please check the merge in `_merge_chunk_attention_outputs()`? The LSE-derived merge weights are currently detached before normalization. Those weights depend on Q/K and are part of the normalized attention merge, so detaching them may change the training gradients relative to the intended attention computation.
> 
> Please remove the detach or justify it with a gradient-parity test.

Thanks for catching this. `_merge_chunk_attention_outputs()` now keeps the LSE-derived merge weights in autograd, and I added both exact float64 merge-gradient tests and direct token-level output/Q/K/V gradient-parity tests.

**Comment by Ternura143** [comment]
@yaoyu-33 I’ve pushed the review updates and refreshed the PR description. The updated implementation includes the transformer-level move, documentation, differentiable LSE merging, CPU/CUDA gradient-parity coverage, and GPT builder/provider smoke validation. Could you please take another look when you have time? Many thanks!
